feat/email verification

shawnchoi8 opened the pull request:
## Summary (요약)
회원가입시 이메일 인증 기능 추가

## Changes (변경 사항)
- 회원가입시 가입한 이메일로 인증 이메일 발송
- 이메일 인증 로직 추가
- 이메일 인증시 발생할 수 있는 예외 추가
- 유저 상태코드 INACTIVE (이메일 인증 전) 추가
- 유저 상태코드 변경 로직 추가 (INACTIVE -> ACTIVE)
- code-data.sql 파일 수정 (INACTIVE 코드 추가)
- 이메일 인증 : GET /api/users/verify-email
- 이메일 재발송 : POST /api/users/resend-verification


## Type of change
- [ ] Bug fix (버그 수정)
- [x] New feature (새로운 기능 추가)
- [x] Refactoring (리팩토링)
- [ ] Documentation (문서화)
- [ ] Tests (테스트 추가/수정)

---

## 📌 PR Checklist
- [x] 브랜치 이름 규칙 확인 (feature/*, fix/* 등)
- [x] PR 대상 브랜치 확인 (develop)
- [x] 최신 develop 반영 완료 (merge or rebase)
- [x] 코드 자동 정렬 실행 (⌥⌘L / Ctrl+Alt+L) & 불필요한 import 제거
- [x] 로컬 빌드/테스트 확인 완료

shawnchoi8 committed:
feat: Add email functionality dependencies and configuration

- Add spring-boot-starter-mail dependency
- Configure Gmail SMTP settings in application.yml
- Set up email username and password in application-personal.yml

shawnchoi8 committed:
feat: Create EmailVerificationToken entity and repository

- Add EmailVerificationToken entity with token, user, and expiration fields
- Implement token expiration and verification status check methods
- Create EmailVerificationTokenRepository with token lookup methods

shawnchoi8 committed:
feat: Implement email verification system

- Add EmailVerificationToken entity with token expiration logic
- Create EmailVerificationTokenRepository for token management
- Implement EmailService with Gmail SMTP integration
- Update signup flow to create INACTIVE users requiring verification
- Add email verification endpoint GET /api/users/verify-email
- Add email resend endpoint POST /api/users/resend-verification
- Send verification emails with clickable links after signup
- Activate user accounts upon successful email verification
- Add EmailVerificationResponseDto for API responses
- Configure mail properties for Gmail SMTP server
- Add Exceptions related to email verification
- fix code-data.sql : add inactive status

soojjung commented on src/main/java/com/iherbyou/user/service/EmailService.java:
```diff
@@ -0,0 +1,88 @@
+            helper.setTo(toEmail); // 회원가입 한 사람의 email (받는 사람 이메일 설정)
+            helper.setSubject("[iHerbYou] [verification email] 이메일 인증을 완료해주세요");
+
+            // 인증 링크 생성
+            String verificationUrl = "http://localhost:8080/api/users/verify-email?token=" + token;
```
여기 `http://localhost:8080`은 환경변수로 빠져야 할 것 같아요

shawnchoi8 replied:
확인했습니다 변경하겠습니다

shawnchoi8 replied:
변경 완료했습니다

soojjung commented on src/main/java/com/iherbyou/user/service/EmailService.java:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * 이메일 인증 메일 HTML 생성
+     */
+    private String buildVerificationEmailHtml(String verificationUrl) {
```
???? 프론트 화면 제가 만들면 되는거죠?

shawnchoi8 replied:
이미 제가 만들어놨습니다 ㅋㅋ 아래가 화면이에요

soojjung commented on src/main/java/com/iherbyou/user/service/UserService.java:
```diff
@@ -48,10 +58,10 @@ public SignUpResponseDto signUp(SignUpRequestDto request) {
-        Code defaultUserStatus = codeService.getDefaultActiveStatus();
-        if (defaultUserStatus == null) {
-            throw new IllegalStateException("기본 사용자 상태를 찾을 수 없습니다");
+        // 기본 사용자 상태 조회 (그룹 71, 코드 711 = INACTIVE) -> DB에 필수 기준 데이터가 있는지 확인하는 안전장치
+        Code inactiveStatus = codeService.getCode(71, 711); // 이메일 인증 전이므로 inactive 상태
+        if (inactiveStatus == null) {
+            throw new IllegalStateException("비활성 상태 코드를 찾을 수 없습니다.");
```
굿굿 👍

soojjung commented on src/main/java/com/iherbyou/user/service/UserService.java:
```diff
@@ -61,23 +71,108 @@ public SignUpResponseDto signUp(SignUpRequestDto request) {
         return SignUpResponseDto.builder()
                 .email(savedUser.getEmail())
-                .message("회원가입이 완료되었습니다. 로그인 해주세요.")
+                .message("회원가입이 완료되었습니다. 이메일을 확인하여 인증을 완료해주세요.")
```
근데 이메일 인증은 사실상 회원가입 도중에 인증하고, 인증되면 가입이 되도록 하는게 낫지 않을까여?

shawnchoi8 replied:
그렇게 진행되고 있는데 message를 좀 고쳐야겠네요

shawnchoi8 replied:
수정 완료했습니다

soojjung commented on src/main/java/com/iherbyou/user/service/UserService.java:
```diff
@@ -61,23 +71,108 @@ public SignUpResponseDto signUp(SignUpRequestDto request) {
+
+        // 사용자 계정 활성화 (INACTIVE -> ACTIVE)
+        User user = verificationToken.getUser();
+        Code activeStatus = codeService.getCode(71, 712);// ACTIVE
```
이 71과 712가 반복적으로 쓰인다면 상수값으로 빠지는건 어떨까여

shawnchoi8 replied:
요거는 쪼 끔만 다음에 해볼게요

soojjung commented on src/main/resources/application.yaml:
```diff
@@ -26,6 +26,22 @@ spring:
+  mail: # 이메일 발송
+    host: smtp.gmail.com
+    port: 587
+    username: ${EMAIL_USERNAME}
+    password: ${EMAIL_PASSWORD}
```
여기 EMAIL_USERNAME과 EMAIL_PASSWORD는 어디서 관리되는 값인가요? 
아마 저한테는 이게 없어서 실행도중에 에러가 발생하는 것 같습니다.

shawnchoi8 replied:
application-personal.yaml 입니다. 지메일 계정이랑 비밀번호가 적혀있어서 personal에 위치해있습니다

soojjung reviewed (commented):
⚽️

soojjung commented on src/main/java/com/iherbyou/user/service/EmailService.java:
```diff
@@ -0,0 +1,88 @@
+     * 이메일 인증 메일 HTML 생성
+     */
+    private String buildVerificationEmailHtml(String verificationUrl) {
+        return """
```
그럼 여기 제가 리액트에 반영시켜 놓을게요 👍

shawnchoi8 replied:
감사합니다

shawnchoi8 committed:
refactor: move email verification URL to configuration fileRetry

- Add app.base-url property to application.yaml
- Inject baseUrl via @Value in EmailService instead of hardcoding
- Enable easy URL changes between local/production environments

shawnchoi8 committed:
chore: modify signup message
